### PR TITLE
feat: render lootrun path on map

### PR DESCRIPTION
**Pull request by HalcyonSuoh**

### CODE REVIEW

**Comment by magicus** [comment]
Ooooh, interesting! O_o 

Can you post a screenshot of how this looks?

**Comment by HalcyonSuoh** [comment]
> Ooooh, interesting! O_o
> 
> Can you post a screenshot of how this looks?

surely
image
image

File: common/src/main/java/com/wynntils/models/lootruns/LootrunCompiler.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/models/lootruns/LootrunCompiler.java]: this should be in the else block I assume

File: common/src/main/java/com/wynntils/models/lootruns/LootrunCompiler.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/models/lootruns/LootrunCompiler.java]: Is this remapping necessary? Why not just store simplified points as Vector2f?

File: common/src/main/java/com/wynntils/utils/render/MapRenderer.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: Shouldn't it be x, 0, z? (Again, does this strictly need to be a Vector3f if we have a filler element?)

File: common/src/main/java/com/wynntils/utils/render/MapRenderer.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/utils/render/MapRenderer.java]:
```suggestion

        Vector3f last = null;
```

File: common/src/main/java/com/wynntils/utils/render/MapRenderer.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: Move to MathUtils perhaps

File: common/src/main/java/com/wynntils/utils/render/MapRenderer.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: Maybe this is worth creating a VectorUtils class

File: common/src/main/java/com/wynntils/utils/render/MapRenderer.java
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: I am more worried about this allocation since this is render time, than the lootrun compiling one.

**Review by kristofbolyai — COMMENTED** [review]
Looks good basically, once we discuss my comments.

File: common/src/main/java/com/wynntils/utils/render/MapRenderer.java
Comment (HalcyonSuoh) [reply common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: > Shouldn't it be x, 0, z?

I honestly don't know, getRenderZ in itself doesn't sounds very intuitive, as it's renderer in screenspace, and z is the forward vector there

Comment (magicus) [reply common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: I agree that this is doing scary much during rendering. All this processing, can't it be done outside rendering? Ideally, we should know exactly what we need to render when it is time to render, and just do that as fast as possible with as few allocations as possible.

Comment (HalcyonSuoh) [reply common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: as long as it still need screenspace transformation it will need to be in render thread, an object pool maybe?

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: > I agree that this is doing scary much during rendering. All this processing, can't it be done outside rendering? Ideally, we should know exactly what we need to render when it is time to render, and just do that as fast as possible with as few allocations as possible.

Perhaps I could profile this. I know it looks like a lot of code, but it is actually rather fast compared to what legacy does for example... Adding more vertexes (or rather not filtering them) is a LOT more slower than this whole process. I will come back here soon with some results.

Comment (Kepler-17c) [reply common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: Legacy didn't optimise at all and just always rendered the whole path, so that's like a worst case or hard limit comparison

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: Yes, I know. But having actual numbers will either make me and magicus relax, or we can find the problem. Win-Win situation.

Comment (Kepler-17c) [reply common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: True - it wasn't my intention to argue against profiling (rather the opposite), just to point out that Legacy isn't a good comparison

File: common/src/main/java/com/wynntils/utils/render/MapRenderer.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: I just noticed. I think we agreed on not using static imports in Artemis (like we don't use import *), but this was not blocked by spotless. Can you refactor?

File: common/src/main/java/com/wynntils/screens/maps/MainMapScreen.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/maps/MainMapScreen.java]: I would at least like to merge these calls, so we don't do most of the work twice. After that, I am most likely fine with performance.

If you don't have time, I can patch it up and merge it. Otherwise, thanks.

File: common/src/main/java/com/wynntils/utils/render/MapRenderer.java
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: Okay. Performance numbers. 97% of the time we spend is actual rendering, around 3% is the math. So I think this concludes the need for optimization.

EDIT: I can't math... the corrected number is: around 90% of time spent rendering. That leaves almost no room to optimize, considering around 9% of the processing time is spent on `Math#atan2`.

EDIT 2: Allocation is a bit more wild... I suggested a "fix" for this, https://github.com/Wynntils/Artemis/pull/1688/files#r1257100736. After that, I think this should be alright.

File: common/src/main/java/com/wynntils/utils/render/MapRenderer.java
Comment (HalcyonSuoh) [reply common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: intellij import it by default, but sure

**Comment by HalcyonSuoh** [comment]
the continuity it's referring to
image

**Comment by kristofbolyai** [comment]
> the continuity it's referring to image

Looks even better!

**Review by kristofbolyai — APPROVED** [review]
This looks great now! 

I know that the code could be simplified a little bit, but I think I (or anyone else) will go ahead and do that later when this is also rendered on the mini-map. While it is only on the main map, the bit higher allocation is acceptable.

**Comment by kristofbolyai** [comment]
As with other features, please wait for 2.0.3 before merging. Thanks.

File: common/src/main/java/com/wynntils/utils/render/MapRenderer.java
Comment (magicus) [reply common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: @Kepler-17c I agree that legacy is not a good comparison. But our bar for Artemis is higher than replicating legacy.

@kristofbolyai Thanks for the measuring. I'm not sure I understand what fix you are pointing to, though...

File: common/src/main/java/com/wynntils/models/lootruns/LootrunCompiler.java
Comment (magicus) [reply common/src/main/java/com/wynntils/models/lootruns/LootrunCompiler.java]: Why was this marked as resolved? I have the same question.

File: common/src/main/java/com/wynntils/utils/render/MapRenderer.java
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: @magicus https://github.com/Wynntils/Artemis/pull/1688#discussion_r1257100736

File: common/src/main/java/com/wynntils/utils/render/MapRenderer.java
Comment (magicus) [inline common/src/main/java/com/wynntils/utils/render/MapRenderer.java]: I'm trying to make sense of this. First you setup a list `points`, then you discard some of them? Can't you combine this, so you only add the points you really need to the `points` list in the first place?

File: common/src/main/java/com/wynntils/models/lootruns/LootrunCompiler.java
Comment (HalcyonSuoh) [reply common/src/main/java/com/wynntils/models/lootruns/LootrunCompiler.java]: >I know that the code could be simplified a little bit, but I think I (or anyone else) will go ahead and do that later when this is also rendered on the mini-map. While it is only on the main map, the bit higher allocation is acceptable.

the reason it was marked as resolved, but I could do it myself I suppose

**Review by kristofbolyai — APPROVED** [review]
Looks better now. Do you think this is fine, @magicus?

**Comment by magicus** [comment]
I would like to take a look, but I can't spend the time rn. Gimme a day or two.

**Review by magicus — APPROVED** [review]
Ship it!

**Pull request merged**
